README.md
# Arduino firmware for the Escornabot Brivoi (DIY version)

This is the Arduino firmware to be uploaded to the **Escornabot Brivoi (DIY version)**.

This is a fork of [the orignal version](https://github.com/escornabot/arduino) tuned by default for [this specific keypad](https://github.com/mgesteiro/escornakeypad) and with some [adittional fixes](https://github.com/escornabot/arduino/pull/18) and small changes too.

More info about the Escornabot at **[escornabot.org](https://escornabot.org)**.


## Install instructions

First of all, you need to donwload and uncompress [this firmware zip file](https://github.com/roboteach-es/escornabot-firmware-DIY/archive/refs/tags/v1.6.2-RT.zip) somewhere in you computer. Then, follow one of these two alternatives:

### Using the Arduino IDE

1. Download and install the [Arduino IDE](https://www.arduino.cc/en/Main/Software).
2. Open the `Escornabot.ino` file in the `Escornabot` subfolder (inside the downloaded compressed file) with the **Arduino IDE**.
3. Choose the appropiate *Board* (=*Arduino Nano*), *Processor* and *Port* (these are specific for you and your computer) in the *Tools* menu, and finally click the upload button.


### Using PlatformIO (advanced)

You'll need to install [PlatformIO CLI](https://platformio.org/install/cli):

    pip install -U platformio

There is already a project file *platformio.ini* to compile and upload the
firmware to the Atmel AVR platform + Arduino framework for the Arduino Nano
and Arduino Micro boards.

#### Arduino Nano

    # compile firmware and upload firmware to an Arduino Nano:
    pio run -e nano -t upload

#### Arduino Micro

    # compile firmware and upload firmware to an Arduino Micro:
    pio run -e micro -t upload


CHANGELOG.md
# Changelog

## v1.6.3-RT (Oct 20, 2021)
New features:
- Support for stepper motors with swapped cables
- Debug option for stepper motors (watch them pulse slowly)
- Human-friendly advance and turning configuration

Fixed bugs:
- RTTTL library parsing/playing

--^^ ROBOteach ^^--------------------------------
--vv ORIGINAL vv---------------------------------

## v1.6.2 (Mar 19, 2019)

New features:
- New "pause" movement, with alternative (long pressed) backwards key.

## v1.6.1 (Jan 5, 2019)

New features:
- New game mode to use 60 degress turns instead 90 (triangled boards).
- Game modes selector with long 'Go' key press.
- Alternative turns (45 and 120 degress) with long key presses.
- Detection of 45 degrees to make the diagonal runs.
- Advise (LED indication) next advance is expected to be a diagonal.
- Beep for long key press is longer than short key press.
- Firmware version is shown as the first message to the serial console.
- Bluetooth parser understands double char as a long key press (alt function).
- Sound indication on game mode changed.
- LEDs on when executing alternative moves.

## v1.6.0 (Jan 5, 2019)

This is an initial v1.6.x release from lastest 1.4.x.

## v1.4.3 (Dec 15, 2017)

Fixed bugs:
- 3-wires keypad not working correctly (reported by Pedro García Pombo).

## v1.4.2 (Oct 28, 2017)

New features:
- EscornaCPU 2.12 configuration file.

## v1.4.1 (Jan 6, 2017)

New features:
- Travis CI is checking pushed code and badge is informing about sanity.
- Refactoring of branches and explained the release numbers in README.
- Nano is set as the default environment in Platformio.

## v1.4.0 (Dec 12, 2016)

New features:
- Standard configurations ready-to-go by board version (see /config folder).
- Instructions to compile and upload with PlatformIO (in README.md).
- Escornabot's greeting preloaded (power on and press the 'Go' button).
- Turn degrees configurable in the abstract Engine class.

## v1.3.3 (Oct 25, 2016)

New features:
- Bluetooth and buttons do not exclude each other as input interface.

## v1.3.2 (Aug 2, 2016)

New features:
- Configuration File to compile with PlatformIO.

Fixed bugs:
- Bluetooth event generation.
- RTTL interpreter default values.

## v1.3.1 (Jun 19, 2016)

New features:
- Leds in keypad are supported.
- Different audio tones by button.
- Enable buzzer by default.
- Default configuration for the [board v2.10](https://github.com/escornabot/electronics/tree/master/Escorna_CPU_2_1) (important: __keypad pin is A7__ instead A4)

Fixed bugs:
- Preprocessor directives reviewed to enable/disable features.
- MoveExecuting event informs the previous move.
- Turn off all LEDs on cancelled program
- RTTL melodies reviewed.



Escornabot/Escornabot.h
// Escornabot.h
/*

Copyright (C) 2014-2019 Escornabot - http://escornabot.com

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
USA.

See LICENSE.txt for details

*/

#ifndef _ESCORNABOT_H
#define _ESCORNABOT_H

#define FIRMWARE_VERSION "v1.6.3-RT"

#include <Arduino.h>
#include "Configuration.h"
#include "Enums.h"
#include "MoveList.h"
#include "Engine.h"
#include "ButtonSet.h"
#include "EventManager.h"

#if defined(ENGINE_TYPE_STEPPERS)

    #include "EngineSteppers.h"
    const EngineSteppers::Config ENGINE_CONFIG = {
        motor_left_in1: STEPPERS_MOTOR_LEFT_IN1,
        motor_left_in2: STEPPERS_MOTOR_LEFT_IN2,
        motor_left_in3: STEPPERS_MOTOR_LEFT_IN3,
        motor_left_in4: STEPPERS_MOTOR_LEFT_IN4,
        motor_right_in1: STEPPERS_MOTOR_RIGHT_IN1,
        motor_right_in2: STEPPERS_MOTOR_RIGHT_IN2,
        motor_right_in3: STEPPERS_MOTOR_RIGHT_IN3,
        motor_right_in4: STEPPERS_MOTOR_RIGHT_IN4,
        steps_per_second: STEPPERS_STEPS_PER_SECOND,
        line_steps: STEPPERS_LINE_STEPS,
        turn_steps: STEPPERS_TURN_STEPS,
    };
    EngineSteppers ENGINE_INSTANCE (&ENGINE_CONFIG);
    Engine* ENGINE = (Engine*) &ENGINE_INSTANCE;

#endif

#if defined(BUTTONS_DIGITAL)

    // digital button set
    #include "ButtonSetDigital.h"
    const ButtonSetDigital::Config BS_CONFIG = {
        pin_button_up: BS_DIGITAL_UP,
        pin_button_right: BS_DIGITAL_RIGHT,
        pin_button_down: BS_DIGITAL_DOWN,
        pin_button_left: BS_DIGITAL_LEFT,
        pin_button_go: BS_DIGITAL_GO,
        pin_button_reset: BS_DIGITAL_RESET,
    };
    ButtonSetDigital BUTTONS_INSTANCE (&BS_CONFIG);
    ButtonSet* BUTTONS = (ButtonSet*) &BUTTONS_INSTANCE;
    #define USE_BUTTONS true

#elif defined(BUTTONS_ANALOG)

    // analog button set
    #include "ButtonSetAnalog.h"
    const ButtonSetAnalog::Config BS_CONFIG = {
        pin_button_set: BS_ANALOG_PIN,
        pullup: (BS_ANALOG_WIRES == 3 ? false : true),
        value_button_up: BS_ANALOG_VALUE_UP,
        value_button_right: BS_ANALOG_VALUE_RIGHT,
        value_button_down: BS_ANALOG_VALUE_DOWN,
        value_button_left: BS_ANALOG_VALUE_LEFT,
        value_button_go: BS_ANALOG_VALUE_GO,
        value_button_reset: BS_ANALOG_VALUE_RESET,
    };
    ButtonSetAnalog BUTTONS_INSTANCE (&BS_CONFIG);
    ButtonSet* BUTTONS = (ButtonSet*) &BUTTONS_INSTANCE;
    #define USE_BUTTONS true

#endif // Button set

#if defined(USE_BLUETOOTH)

    #ifndef BLUETOOTH_SERIAL
        #define BLUETOOTH_SERIAL Serial
    #endif

    #include "BluetoothInterface.h"
    const BluetoothInterface::Config BT_CONFIG = {
        serial: &(BLUETOOTH_SERIAL),
        bauds: BLUETOOTH_BAUDS,
    };
    BluetoothInterface BLUETOOTH_INSTANCE (&BT_CONFIG);
    BluetoothInterface* BLUETOOTH = &BLUETOOTH_INSTANCE;

#endif // Bluetooth


#if USE_BUZZER
    #include "Buzzer.h"
    Buzzer BUZZER = Buzzer(BUZZER_PIN);
#endif


#if USE_SIMPLE_LED
    #include "SimpleLed.h"
    SimpleLed SIMPLE_LED = SimpleLed(SIMPLE_LED_PIN);
#endif


#if USE_KEYPAD_LEDS
    #include "KeypadLeds.h"
    const KeypadLeds::Config KEYPAD_LEDS_CONFIG = {
        pin_led_up: KEYPAD_LED_PIN_UP,
        pin_led_right: KEYPAD_LED_PIN_RIGHT,
        pin_led_down: KEYPAD_LED_PIN_DOWN,
        pin_led_left: KEYPAD_LED_PIN_LEFT,
        pin_led_go: KEYPAD_LED_PIN_GO,
    };
    KeypadLeds KEYPAD_LEDS = KeypadLeds(&KEYPAD_LEDS_CONFIG);
#endif

///// global vars

// event manager
EventManager EVENTS_INSTANCE;
EventManager* EVENTS = &EVENTS_INSTANCE;

// program
MoveList PROGRAM_INSTANCE;
MoveList* PROGRAM = &PROGRAM_INSTANCE;


#endif // _ESCORNABOT_H

// EOF


Escornabot/Configuration.h
// Configuration.h
/*

Copyright (C) 2014-2019 Escornabot - http://escornabot.com

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
USA.

See LICENSE.txt for details

*/

#ifndef _CONFIGURATION_H
#define _CONFIGURATION_H


//////////////////////////////////////////////////////////////////////
///// general configuration
//////////////////////////////////////////////////////////////////////

// engine to use
#define ENGINE_TYPE_STEPPERS
#define STEPPERS_SWAPPED_CABLES false
#define STEPPERS_DEBUG_COILS false

// button set to use (analog input, digital input)
#define BUTTONS_ANALOG
//#define BUTTONS_DIGITAL

// milliseconds after a button is considered as pressed
#define BUTTON_MIN_PRESSED 30

// milliseconds after a button is considered as long pressed
#define BUTTON_LONG_PRESSED 800

// put to false to add movements to the program after its execution
#define PROGRAM_RESET_ALWAYS true

// store configuration and program within internal EEPROM
#define USE_PERSISTENT_MEMORY false

// memory capacity for program movements
#define MOVE_LIMIT 100

// milliseconds for the "pause" movement
#define PAUSE_MOVE_MILLIS 1000

// milliseconds delay before starting to move
#define DELAY_BEFORE_GO 500

// milliseconds to pause after every movement
#define AFTER_MOVEMENT_PAUSE 0

// point of view set when Vacalourabot is started
#define POV_INITIAL    POV_ESCORNABOT

// bluetooth serial
#define USE_BLUETOOTH false
#define BLUETOOTH_BAUDS 9600

// buzzer
#define USE_BUZZER true
#define BUZZER_PIN 10
#define PROGRAM_FINISHED_RTTL RTTL_FIDO
#define TONE_FREQ_UP 2637
#define TONE_FREQ_RIGHT 4434
#define TONE_FREQ_DOWN 3520
#define TONE_FREQ_LEFT 2217

// simple led
#define USE_SIMPLE_LED true
#define SIMPLE_LED_PIN 13

// keypad leds
#define USE_KEYPAD_LEDS false
#define KEYPAD_LED_PIN_UP A0
#define KEYPAD_LED_PIN_RIGHT A3
#define KEYPAD_LED_PIN_DOWN A2
#define KEYPAD_LED_PIN_LEFT A1
#define KEYPAD_LED_PIN_GO 13


//////////////////////////////////////////////////////////////////////
///// Steppers engine setup
//////////////////////////////////////////////////////////////////////

#ifdef ENGINE_TYPE_STEPPERS

// stepper pin setup (digital outputs)
#if !STEPPERS_SWAPPED_CABLES

// STANDARD: YELLOW - ORANGE - RED - PINK - BLUE
#define STEPPERS_MOTOR_RIGHT_IN1 5
#define STEPPERS_MOTOR_RIGHT_IN2 4
#define STEPPERS_MOTOR_RIGHT_IN3 3
#define STEPPERS_MOTOR_RIGHT_IN4 2
#define STEPPERS_MOTOR_LEFT_IN1 9
#define STEPPERS_MOTOR_LEFT_IN2 8
#define STEPPERS_MOTOR_LEFT_IN3 7
#define STEPPERS_MOTOR_LEFT_IN4 6

#else

// SWAPPED: ORANGE - YELLOW - RED - BLUE - PINK
#define STEPPERS_MOTOR_RIGHT_IN1 4
#define STEPPERS_MOTOR_RIGHT_IN2 5
#define STEPPERS_MOTOR_RIGHT_IN3 2
#define STEPPERS_MOTOR_RIGHT_IN4 3
#define STEPPERS_MOTOR_LEFT_IN1 8
#define STEPPERS_MOTOR_LEFT_IN2 9
#define STEPPERS_MOTOR_LEFT_IN3 6
#define STEPPERS_MOTOR_LEFT_IN4 7

#endif

// step calibration
#define WHEEL_RADIUS 38.0 //mm Brivoi DIY
#define MOVE_DISTANCE_DEFAULT 100.0 //mm
#define TURN_DEGREES_DEFAULT 90.0 //°
#define STEPPERMOTOR_FULLREVOLUTION_STEPS 4096 //number of steps for a full revolution of the axis
#define WHEEL_CIRCUNFERENCE float(2.0 * PI * WHEEL_RADIUS)

#define STEPPERS_LINE_STEPS int16_t(STEPPERMOTOR_FULLREVOLUTION_STEPS / WHEEL_CIRCUNFERENCE * MOVE_DISTANCE_DEFAULT)
#define STEPPERS_TURN_STEPS int16_t(STEPPERMOTOR_FULLREVOLUTION_STEPS / 360.0 * TURN_DEGREES_DEFAULT)
#define STEPPERS_STEPS_PER_SECOND 1000 // 1500 max. with fully charged batteries

#endif



//////////////////////////////////////////////////////////////////////
///// Button set digital
//////////////////////////////////////////////////////////////////////

#ifdef BUTTONS_DIGITAL

// keypad pin setup (digital or analog inputs) (use 255 if key doesn't exist)
#define BS_DIGITAL_UP A0
#define BS_DIGITAL_RIGHT A1
#define BS_DIGITAL_DOWN A2
#define BS_DIGITAL_LEFT A3
#define BS_DIGITAL_GO A4
#define BS_DIGITAL_RESET 255

#endif // BUTTONS_DIGITAL



//////////////////////////////////////////////////////////////////////
///// Button set analog
//////////////////////////////////////////////////////////////////////

#ifdef BUTTONS_ANALOG

#define BS_ANALOG_WIRES 2
//#define BS_ANALOG_WIRES 3

// keypad pin setup (analog input)
#define BS_ANALOG_PIN A4

// input values for each key pressed (0 if key doesn't exist)
// values pre-configured for this keypad -> https://github.com/mgesteiro/escornakeypad
#define BS_ANALOG_VALUE_UP 220
#define BS_ANALOG_VALUE_RIGHT 510
#define BS_ANALOG_VALUE_DOWN 440
#define BS_ANALOG_VALUE_LEFT 350
#define BS_ANALOG_VALUE_GO 615
#define BS_ANALOG_VALUE_RESET 0

#endif // BUTTONS_ANALOG



//////////////////////////////////////////////////////////////////////
///// Button set Bluetooth
//////////////////////////////////////////////////////////////////////

#ifdef USE_BLUETOOTH

// Arduino serial port (default is Serial, use Serial1 with Arduino Micro)
//#define BLUETOOTH_SERIAL    Serial
//#define BLUETOOTH_SERIAL    Serial1
//#define BLUETOOTH_SERIAL    Serial2
//#define BLUETOOTH_SERIAL    Serial3

#endif // USE_BLUETOOTH




//////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////


#endif // _CONFIGURATION_H

// EOF


Escornabot/EngineSteppers.cpp
// EngineSteppers.cpp
/*

Copyright (C) 2014-2019 Escornabot - http://escornabot.com

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
USA.

See LICENSE.txt for details

*/

#include "EngineSteppers.h"
#include "EventManager.h"
#include <Arduino.h>

//////////////////////////////////////////////////////////////////////

const static uint8_t step_pattern[] = {
    B00001, B00011, B00010, B00110, B00100, B01100, B01000, B01001
};

//////////////////////////////////////////////////////////////////////

extern EventManager* EVENTS;

//////////////////////////////////////////////////////////////////////

EngineSteppers::EngineSteppers(const Config* cfg)
{
    _config = cfg;

    _movement_steps_r = 0;
    _movement_steps_l = 0;

    _pattern_index_left = 0;
    _pattern_index_right = 0;
}

//////////////////////////////////////////////////////////////////////

void EngineSteppers::init()
{
    pinMode(_config->motor_left_in1, OUTPUT);
    pinMode(_config->motor_left_in2, OUTPUT);
    pinMode(_config->motor_left_in3, OUTPUT);
    pinMode(_config->motor_left_in4, OUTPUT);
    pinMode(_config->motor_right_in1, OUTPUT);
    pinMode(_config->motor_right_in2, OUTPUT);
    pinMode(_config->motor_right_in3, OUTPUT);
    pinMode(_config->motor_right_in4, OUTPUT);

    // set coils in row
    _movement_steps_r = 8;
    _movement_steps_l = 8;

    EVENTS->add(this);
}

//////////////////////////////////////////////////////////////////////

void EngineSteppers::turn(int16_t degrees)
{
    int32_t steps = -_config->turn_steps;
    steps *= degrees;
    steps /= 90;

    _movement_steps_r = steps;
    _movement_steps_l = steps;
}

//////////////////////////////////////////////////////////////////////

void EngineSteppers::moveStraight(float advance_units)
{
    _movement_steps_r = _config->line_steps * advance_units;
    _movement_steps_l = -_movement_steps_r;
}

//////////////////////////////////////////////////////////////////////

void EngineSteppers::_motorStepRight(uint8_t pattern)
{
    digitalWrite(_config->motor_right_in1, bitRead(pattern, 0));
    digitalWrite(_config->motor_right_in2, bitRead(pattern, 1));
    digitalWrite(_config->motor_right_in3, bitRead(pattern, 2));
    digitalWrite(_config->motor_right_in4, bitRead(pattern, 3));
}

//////////////////////////////////////////////////////////////////////

void EngineSteppers::_motorStepLeft(uint8_t pattern)
{
    digitalWrite(_config->motor_left_in1, bitRead(pattern, 0));
    digitalWrite(_config->motor_left_in2, bitRead(pattern, 1));
    digitalWrite(_config->motor_left_in3, bitRead(pattern, 2));
    digitalWrite(_config->motor_left_in4, bitRead(pattern, 3));
}

//////////////////////////////////////////////////////////////////////

void EngineSteppers::tick(uint32_t micros)
{
    if (!isExecuting()) return;

    if (_movement_steps_l == 0 && _movement_steps_r == 0)
    {
        // already executing a pause movement
        if (_inPauseMove()) return;

        // deactivate Darlingtons
        _motorStepLeft(0);
        _motorStepRight(0);

        EVENTS->indicateMoveExecuted(_getCurrentMove());

        // prepare next movement
        _program_index++;
        _prepareMove();
        return;
    }

    if (_movement_steps_l != 0)
    {
        // left motor step
        int8_t delta = (_movement_steps_l > 0 ? 1 : -1);
        _motorStepLeft(step_pattern[_pattern_index_left]);
        _pattern_index_left += delta + 8;
        _pattern_index_left %= 8;
        _movement_steps_l -= delta;
    }

    if (_movement_steps_r != 0)
    {
        // right motor step
        int8_t delta = (_movement_steps_r > 0 ? 1 : -1);
        _motorStepRight(step_pattern[_pattern_index_right]);
        _pattern_index_right += delta + 8;
        _pattern_index_right %= 8;
        _movement_steps_r -= delta;
    }

    if (_is_cancelling)
    {
        _movement_steps_r = 0;
        _movement_steps_l = 0;
    }
    else
    {
        #if !STEPPERS_DEBUG_COILS
        delayMicroseconds(1000000 / _config->steps_per_second);
        #else
        delay(1000);
        #endif
    }
}

//////////////////////////////////////////////////////////////////////

// EOF


Escornabot/Bot.cpp
// Bot.cpp
/*

Copyright (C) 2014-2019 Escornabot - http://escornabot.com

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
USA.

See LICENSE.txt for details

*/

#include "Bot.h"
#include "Escornabot.h"

//////////////////////////////////////////////////////////////////////

void Bot::init()
{
    // init engine
    ENGINE->init();

    // init button set
    #if USE_BUTTONS
    BUTTONS->init();
    #endif

    // early console
    Serial.begin(9600);
    Serial.println(FIRMWARE_VERSION);
    delay(100);

    // init bluetooth
    #if USE_BLUETOOTH
    BLUETOOTH->init();
    #endif

    // init buzzer
    #if USE_BUZZER
    BUZZER.init();
    BUZZER.playRttl(RTTL_STARTUP);
    #endif

    #if USE_SIMPLE_LED
    SIMPLE_LED.init();
    #endif

    #if USE_KEYPAD_LEDS
    KEYPAD_LEDS.init();
    #endif

    // restore last program
    #if USE_PERSISTENT_MEMORY
    PROGRAM->load();
    #endif

    EVENTS->add(this);
}

//////////////////////////////////////////////////////////////////////

void Bot::loop()
{
    EVENTS->indicateTick(micros());

    if (!ENGINE->isExecuting()) delay(10);
}

//////////////////////////////////////////////////////////////////////

void Bot::buttonPressed(BUTTON button)
{
    #if USE_SIMPLE_LED
    SIMPLE_LED.setStatus(true);
    #endif
}

//////////////////////////////////////////////////////////////////////

void Bot::buttonReleased(BUTTON button)
{
    #if USE_SIMPLE_LED
    SIMPLE_LED.setStatus(false);
    #endif

    if (ENGINE->isExecuting())
    {
        ENGINE->cancelExecution();
        return;
    }

    switch (button)
    {
        case BUTTON_UP:
            _storeMove(MOVE_FORWARD);
            break;

        case BUTTON_RIGHT:
            _storeMove(MOVE_RIGHT);
            break;

        case BUTTON_DOWN:
            _storeMove(MOVE_BACKWARD);
            break;

        case BUTTON_LEFT:
            _storeMove(MOVE_LEFT);
            break;

        case BUTTON_GO:
            _go();
            break;

        case BUTTON_RESET:
            PROGRAM->clear();
            break;
    }
}

//////////////////////////////////////////////////////////////////////

void Bot::buttonLongReleased(BUTTON button)
{
    #if USE_SIMPLE_LED
    SIMPLE_LED.setStatus(false);
    #endif

    if (ENGINE->isExecuting())
    {
        ENGINE->cancelExecution();
        return;
    }

    switch (button)
    {
        case BUTTON_RIGHT:
            _storeMove(MOVE_ALT_RIGHT);
            break;

        case BUTTON_DOWN:
            _storeMove(MOVE_PAUSE);
            break;

        case BUTTON_LEFT:
            _storeMove(MOVE_ALT_LEFT);
            break;

        case BUTTON_GO:
            _next_game_mode();
            break;
    }
}

//////////////////////////////////////////////////////////////////////

void Bot::programFinished()
{
    #if PROGRAM_RESET_ALWAYS
    PROGRAM->clear();
    #endif

    #if USE_BUZZER
    BUZZER.playRttl(PROGRAM_FINISHED_RTTL);
    #endif

    if (_game_mode == GAME_MODE_GRID_90 && !ENGINE->isAligned(90))
    {
        // advise next advance is expected to be a diagonal:
        #if USE_SIMPLE_LED
        SIMPLE_LED.setStatus(true);
        #elif USE_KEYPAD_LEDS
        KEYPAD_LEDS.setLed(BUTTON_RIGHT, true);
        KEYPAD_LEDS.setLed(BUTTON_LEFT, true);
        #endif
    }
}

//////////////////////////////////////////////////////////////////////

void Bot::programAborted(uint8_t executed, uint8_t total)
{
    #if PROGRAM_RESET_ALWAYS
    PROGRAM->clear();
    #endif
}

//////////////////////////////////////////////////////////////////////

void Bot::_go()
{
    /*
    // preloaded program
    if (PROGRAM->getMoveCount() == 0 && _total_programs == 0)
    {
        PROGRAM->addProgram(PROGRAM_ESCORNA_GREETING);
    }
    */

    // only with movements
    if (PROGRAM->getMoveCount() > 0)
    {
        #if USE_PERSISTENT_MEMORY
        // save the movement program in EEPROM
        PROGRAM->save();
        #endif

        // let user to release the 'go' button before the action
        delay(DELAY_BEFORE_GO);

        ENGINE->execute(PROGRAM, AFTER_MOVEMENT_PAUSE, POV_ESCORNABOT);
        _total_programs++;
    }
}

//////////////////////////////////////////////////////////////////////

void Bot::_storeMove(MOVE move)
{
    if (PROGRAM->getMoveCount() < MOVE_LIMIT)
    {
        PROGRAM->addMove(move);
    }
    else
    {
        // memory is full
        _go();
    }
}

//////////////////////////////////////////////////////////////////////

void Bot::_next_game_mode()
{
    // select the next game mode (2 modes currently available)
    ++_game_mode %= 2;

    switch (_game_mode)
    {
        case GAME_MODE_GRID_90:
            PROGRAM->setTurnDegrees(90);
            PROGRAM->setAltTurnDegrees(45);
            ENGINE->setSquareDiagonals(true);
            break;

        case GAME_MODE_GRID_60:
            PROGRAM->setTurnDegrees(60);
            PROGRAM->setAltTurnDegrees(120);
            ENGINE->setSquareDiagonals(false);
            break;
    }

    EVENTS->indicateGameModeSelected(_game_mode);
}

//////////////////////////////////////////////////////////////////////

// EOF
